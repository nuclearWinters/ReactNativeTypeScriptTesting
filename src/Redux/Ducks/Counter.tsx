import {RootState} from '../Reducers/Reducers';
import {ThunkAction} from 'redux-thunk';
/* Reducer */
const initialState: CounterState = {
  integer: 0,
  string: 'Â¡Cambia el estado en Redux!',
};

interface CounterState {
  integer: number;
  string: string;
}

/* TYPES */
const DECREASE_COUNTER = 'DECREASE_COUNTER';
const INCREASE_COUNTER = 'INCREASE_COUNTER';
const SET_COUNTER = 'SET_COUNTER';
const SET_STRING = 'SET_STRING';

interface IncreaseCounter {
  type: typeof INCREASE_COUNTER;
}

interface DecreaseCounter {
  type: typeof DECREASE_COUNTER;
}

interface SetCounter {
  type: typeof SET_COUNTER;
  payload: number;
}

interface SetString {
  type: typeof SET_STRING;
  payload: string;
}

export type CounterActionTypes =
  | DecreaseCounter
  | IncreaseCounter
  | SetCounter
  | SetString;

const decreaseCounter = (): DecreaseCounter => ({
  type: DECREASE_COUNTER,
});

const increaseCounter = (): IncreaseCounter => ({
  type: INCREASE_COUNTER,
});

const setCounter = (payload: number): SetCounter => ({
  type: SET_COUNTER,
  payload,
});

const setString = (payload: string): SetString => ({
  type: SET_STRING,
  payload,
});

const awaitedSetTimeout2 = new Promise<number>((resolve) => {
  setTimeout(() => {
    resolve(10);
  }, 5000);
});

const mockFetch = (): ThunkAction<
  Promise<number>,
  RootState,
  undefined,
  CounterActionTypes
> => {
  return async (dispatch, getState) => {
    let {
      counter: {integer},
    } = getState();
    try {
      const response = await awaitedSetTimeout2;
      dispatch(setCounter(integer + 10));
      return response;
    } catch (e) {
      return e;
    }
  };
};

export const reducer = (
  state = initialState,
  action: CounterActionTypes,
): CounterState => {
  switch (action.type) {
    case DECREASE_COUNTER:
      return {...state, integer: state.integer - 1};
    case INCREASE_COUNTER:
      return {...state, integer: state.integer + 1};
    case SET_COUNTER:
      return {...state, integer: action.payload};
    case SET_STRING:
      return {...state, string: action.payload};
    default:
      return state;
  }
};

export {decreaseCounter, increaseCounter, setCounter, setString, mockFetch};
