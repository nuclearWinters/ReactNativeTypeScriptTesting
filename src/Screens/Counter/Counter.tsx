// @refresh reset
import React from 'react';
import {View, Text} from 'react-native';
import {useDispatch} from 'react-redux';
import {useTypedSelector, RootState} from '../../Redux/Reducers/Reducers';
import {RootStackParamList} from '../../../App';
import {Counter} from '../../Redux/Ducks';
import {ThunkDispatch} from 'redux-thunk';
import {CounterActionTypes} from '../../Redux/Ducks/Counter';
import BorderButton from '../../Library/BaseButton/BorderButton';
import WideButton from '../../Library/BaseButton/WideButton';
import AnimatedInput from '../../Library/BaseInput/AnimatedInput';
import {StackScreenProps} from '@react-navigation/stack';

type Props = StackScreenProps<RootStackParamList, 'Counter'>;

const CounterScreen: React.FC<Props> = (props) => {
  const {route} = props;
  const counter = useTypedSelector((state) => state.counter);
  const dispatch = useDispatch<
    ThunkDispatch<RootState, undefined, CounterActionTypes>
  >();
  return (
    <View>
      <Text style={{textAlign: 'center', fontSize: 32, paddingVertical: 14}}>
        {route.params.itemId}
      </Text>
      <Text style={{textAlign: 'center', marginVertical: 20, fontSize: 24}}>
        {counter.string}
      </Text>
      <View style={{flexDirection: 'row'}}>
        <BorderButton
          text={'Menos'}
          onPress={() => {
            dispatch(Counter.decreaseCounter());
          }}
        />
        <BorderButton
          text={'Reiniciar'}
          onPress={() => {
            dispatch(Counter.setCounter(route.params.itemId));
          }}
        />
        <BorderButton
          text={'Más'}
          onPress={() => {
            dispatch(Counter.increaseCounter());
          }}
        />
      </View>
      <Text style={{textAlign: 'center', fontSize: 32, paddingVertical: 14}}>
        {counter.integer}
      </Text>
      <AnimatedInput
        placeholder="Ingresa un nuevo título"
        value={counter.string}
        onChangeText={(text) => {
          dispatch(Counter.setString(text));
        }}
        placeholderColor="gray"
      />
      <WideButton
        text="PRESS ME"
        onPress={() => {
          dispatch(Counter.mockFetch())
            .then((res) => {
              console.log(res);
            })
            .catch((e) => {
              console.log(e);
            });
        }}
      />
    </View>
  );
};

export default CounterScreen;
