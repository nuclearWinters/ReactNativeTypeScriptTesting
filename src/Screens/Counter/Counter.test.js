import React from 'react';
import {render, fireEvent, cleanup} from '@testing-library/react-native';
import Counter from './Counter';
import {Provider} from 'react-redux';
import store from '../../Redux/Store';

afterEach(cleanup);

test('Home', async () => {
  const {getByText, baseElement} = render(
    <Provider store={store}>
      <Counter route={{params: {itemId: 10}}} />
    </Provider>,
  );
  const title1 = getByText('10');
  expect(title1).toBeTruthy();
  const title2 = getByText('¡Cambia el estado en Redux!');
  expect(title2).toBeTruthy();
  const button1 = getByText('Menos');
  expect(button1).toBeTruthy();
  const button2 = getByText('Reiniciar');
  expect(button2).toBeTruthy();
  const button3 = getByText('Más');
  expect(button3).toBeTruthy();
  const counter = getByText('0');
  expect(counter).toBeTruthy();
  const button4 = getByText('PRESS ME');
  expect(button4).toBeTruthy();
  fireEvent.press(button1);
  expect(counter.props.children).toBe(-1);
  expect(baseElement).toMatchSnapshot();
});
