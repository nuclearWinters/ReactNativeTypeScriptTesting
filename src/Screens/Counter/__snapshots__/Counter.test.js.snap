// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "fontSize": 32,
            "paddingVertical": 14,
            "textAlign": "center",
          }
        }
      >
        10
      </Text>
      <Text
        style={
          Object {
            "fontSize": 24,
            "marginVertical": 20,
            "textAlign": "center",
          }
        }
      >
        ¡Cambia el estado en Redux!
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <TouchableOpacity
          testID="prueba"
        />
        <TouchableOpacity
          style={
            Array [
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ],
            ]
          }
          testID="box"
        >
          <Text
            style={
              Array [
                Array [
                  Object {
                    "borderRadius": 6,
                    "borderWidth": 1,
                    "fontSize": 18,
                    "paddingHorizontal": 18,
                    "paddingVertical": 8,
                  },
                  undefined,
                ],
              ]
            }
            testID="text"
          >
            Menos
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={
            Array [
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ],
            ]
          }
          testID="box"
        >
          <Text
            style={
              Array [
                Array [
                  Object {
                    "borderRadius": 6,
                    "borderWidth": 1,
                    "fontSize": 18,
                    "paddingHorizontal": 18,
                    "paddingVertical": 8,
                  },
                  undefined,
                ],
              ]
            }
            testID="text"
          >
            Reiniciar
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={
            Array [
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ],
            ]
          }
          testID="box"
        >
          <Text
            style={
              Array [
                Array [
                  Object {
                    "borderRadius": 6,
                    "borderWidth": 1,
                    "fontSize": 18,
                    "paddingHorizontal": 18,
                    "paddingVertical": 8,
                  },
                  undefined,
                ],
              ]
            }
            testID="text"
          >
            Más
          </Text>
        </TouchableOpacity>
      </View>
      <Text
        style={
          Object {
            "fontSize": 32,
            "paddingVertical": 14,
            "textAlign": "center",
          }
        }
      >
        -1
      </Text>
      <View
        style={
          Array [
            Array [
              Object {
                "marginHorizontal": 40,
              },
              undefined,
            ],
          ]
        }
        testID="box"
      >
        <TextInput
          allowFontScaling={true}
          placeholder="Ingresa un nuevo título"
          placeholderTextColor="gray"
          rejectResponderTermination={true}
          style={
            Array [
              Array [
                Object {
                  "borderBottomColor": "lightgray",
                  "borderBottomWidth": 2,
                  "color": "black",
                  "fontSize": 18,
                },
                undefined,
              ],
            ]
          }
          testID="textInput"
          underlineColorAndroid="transparent"
          value="¡Cambia el estado en Redux!"
        />
        <View
          style={
            Object {
              "backgroundColor": "rgb(0,100,180)",
              "bottom": 0,
              "height": 2,
              "left": "50%",
              "position": "absolute",
              "right": "50%",
            }
          }
          testID="borderBottom"
        />
      </View>
      <TouchableOpacity
        style={
          Array [
            Array [
              Object {
                "alignSelf": "center",
                "backgroundColor": "rgb(0,100,180)",
                "borderRadius": 100,
                "marginVertical": 64,
                "paddingVertical": 12,
                "width": "80%",
              },
              undefined,
            ],
          ]
        }
        testID="box"
      >
        <Text
          style={
            Array [
              Array [
                Object {
                  "color": "white",
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                undefined,
              ],
            ]
          }
          testID="text"
        >
          PRESS ME
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
